### PR TITLE
Update Date Validation

### LINKED ISSUE #208: BUG: Adding txn with non existent date of 30022024 is not thrown error but parsed as 29022024
Reproduce by `addTxn p/98765432 amt/12.3 desc/John paid me for dinner date/30022024`

Screenshot 2024-10-29 at 9 57 31 PM

### PR DESCRIPTION
**Pull request by rollingpencil**
Update Date validation to be strict and check for invalid dates in Feb
Resolves #208

### CODE REVIEW

File: src/test/java/spleetwaise/transaction/model/transaction/DateTest.java [hunk 1]
```diff
@@ -31,6 +31,8 @@ public void isValidDate_null_exceptionThrown() {
     @Test
     public void isValidDate_invalidInput_returnsFalse() {
         assertFalse(Date.isValidDate(""));
+        assertFalse(Date.isValidDate("     "));
+        assertFalse(Date.isValidDate("30022024")); // Feb edge case
```
Comment (SeeYangZhi) [inline src/test/java/spleetwaise/transaction/model/transaction/DateTest.java:35]: Recommend to refer to week 11 tutorial recording to add test cases based on the date equivalence partion: [1-28],[29],[30],[31]. Some years's feb accept 29 as the day, some months dont accept 31.

File: src/main/java/spleetwaise/transaction/model/transaction/Date.java [hunk 2]
```diff
@@ -15,9 +16,10 @@
 public class Date {
     public static final String MESSAGE_CONSTRAINTS = "Date should only in the format of DDMMYYYY";
 
-    public static final DateTimeFormatter VALIDATION_FORMATTER = DateTimeFormatter.ofPattern("ddMMyyyy");
+    public static final DateTimeFormatter VALIDATION_FORMATTER = DateTimeFormatter.ofPattern("ddMMuuuu")
```
Comment (SeeYangZhi) [inline src/main/java/spleetwaise/transaction/model/transaction/Date.java:19]: Just want to learn, what is the reason for the change to uuuu?

**Review by SeeYangZhi — CHANGES_REQUESTED** [review]

File: src/main/java/spleetwaise/transaction/model/transaction/Date.java (hunk 2 above)
Comment (rollingpencil) [reply src/main/java/spleetwaise/transaction/model/transaction/Date.java:19]: due to the strict validation rules, `yyyy` refers to year-of-era and `uuuu` refers to year. Apparently `2024` is valid for `uuuu` but not `yyyy`

File: src/test/java/spleetwaise/transaction/model/transaction/DateTest.java (hunk 1 above)
Comment (rollingpencil) [reply src/test/java/spleetwaise/transaction/model/transaction/DateTest.java:35]: In this case, the validation might be harder since we are validating by the whole date rather than just the day fields.

Comment (SeeYangZhi) [reply src/test/java/spleetwaise/transaction/model/transaction/DateTest.java:35]: I see, could u point us to the documentation?

File: src/main/java/spleetwaise/transaction/model/transaction/Date.java (hunk 2 above)
Comment (SeeYangZhi) [reply src/main/java/spleetwaise/transaction/model/transaction/Date.java:19]: I see, could you point us to the documentation?

File: src/test/java/spleetwaise/transaction/model/transaction/DateTest.java (hunk 1 above)
Comment (SeeYangZhi) [reply src/test/java/spleetwaise/transaction/model/transaction/DateTest.java:35]: You mean the current implementation still allows invalid date? Or the test cases needs more work

File: src/main/java/spleetwaise/transaction/model/transaction/Date.java (hunk 2 above)
Comment (rollingpencil) [reply src/main/java/spleetwaise/transaction/model/transaction/Date.java:19]: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/time/format/DateTimeFormatter.html

Screenshot 2024-11-04 at 4 30 45 PM

Initially though `yyyy` would work but with the strict requirement only `uuuu` works

File: src/main/java/spleetwaise/transaction/model/transaction/Date.java [hunk 3]
```diff
@@ -15,9 +16,10 @@
 public class Date {
     public static final String MESSAGE_CONSTRAINTS = "Date should only in the format of DDMMYYYY";
```
Comment (SeeYangZhi) [inline src/main/java/spleetwaise/transaction/model/transaction/Date.java:17]: Should we change this to reflect that it should only be a valid date since iirc this is also used as an error message?

**Review by SeeYangZhi — CHANGES_REQUESTED** [review]
